backend/aopp.go

```diff
@@ -0,0 +1,303 @@
+func (backend *Backend) aoppSetError(err ErrorCode) {
+	backend.aopp.State = aoppStateError
+	backend.aopp.ErrorCode = err
+	backend.notifyAOPP()
+}
+
+// aoppKeystoreRegistered must be called after a keystore is available, to display a list of
+// accounts to choose from. `accountsAndKeystoreLock` must be held when calling this function.
+func (backend *Backend) aoppKeystoreRegistered() {
+	if backend.aopp.State != aoppStateAwaitingKeystore {
+		return
+	}
+	if !backend.keystore.CanSignMessage(backend.aopp.coinCode) {
+		backend.aoppSetError(errAoppCannotSign)
+		return
+	}
+	var accounts []account
+	for _, acct := range backend.accounts {
+		if !acct.Config().Active {
+			continue
+		}
+		if acct.Coin().Code() != backend.aopp.coinCode {
+			continue
+		}
+		accounts = append(accounts, account{
+			Name: acct.Config().Name,
+			Code: acct.Config().Code,
+		})
+	}
+
+	if len(accounts) == 0 {
+		backend.aoppSetError(errAoppNoAccounts)
+		return
+	}
+
+	backend.aopp.Accounts = accounts
+	backend.aopp.State = aoppStateChoosingAccount
+	backend.notifyAOPP()
+}
+
+// handleAOPP handles an AOPP (Address Ownership Proof Protocol) request. See https://aopp.group/.
+func (backend *Backend) handleAOPP(uri *url.URL) {
+	defer backend.accountsAndKeystoreLock.Lock()()
+
+	log := backend.log.WithField("aopp-uri", uri.String())
+
+	q := uri.Query()
+
+	backend.aopp.CallbackHost = "<unknown>"
+
+	if q.Get("v") != "0" {
+		log.Error("Can only handle version 0 aopp URIs")
+		backend.aoppSetError(errAoppVersion)
+		return
+	}
+
+	callback := q.Get("callback")
+	if callback == "" {
+		log.Error("callback param missing")
+		return
+	}
+	callbackURL, err := url.Parse(callback)
+	if err != nil {
+		log.WithError(err).Error("Invalid callback")
+		backend.aoppSetError(errAoppInvalidRequest)
+		return
+	}
+	backend.aopp.CallbackHost = callbackURL.Host
+	backend.aopp.callback = callback
+
+	coinCode, ok := map[string]coinpkg.Code{
+		"btc": coinpkg.CodeBTC,
+		// TODO: add support for ETH
+		// "eth": coinpkg.CodeETH,
+	}[strings.ToLower(q.Get("asset"))]
+	if !ok {
+		log.Error("Unrecognized coin")
+		backend.aoppSetError(errAoppUnsupportedAsset)
+		return
+	}
+	backend.aopp.coinCode = coinCode
+
+	msg := q.Get("msg")
+	if msg == "" {
+		log.Error("msg param missing")
+		backend.aoppSetError(errAoppInvalidRequest)
+		return
+	}
+	backend.aopp.message = msg
+
+	backend.aopp.State = aoppStateAwaitingKeystore
+	if backend.keystore == nil {
+		backend.notifyAOPP()
+		return
+	}
+	backend.aoppKeystoreRegistered()
+}
+
+// AOPPChooseAccount is called when an AOPP request is being processed and the user has chosen an
+// account.
+func (backend *Backend) AOPPChooseAccount(code accounts.Code) {
+	defer backend.accountsAndKeystoreLock.Lock()()
+	if backend.aopp.State != aoppStateChoosingAccount {
+		return
+	}
+
+	backend.aopp.State = aoppStateSyncing
+	backend.notifyAOPP()
+
+	log := backend.log.WithField("accountCode", code)
+	account := func() accounts.Interface {
```

Comment:
don't really need a closure func. it gets shorter:

```go
var account accounts.Interface
for _, acct := range backend.accounts {
    if acct.Config().Code == code {
      account = acct
    }
}
// account is nil if not found
```

Reply:
:heavy_check_mark: